src/libmongoc/src/mongoc/mongoc-server-monitor.c

```diff
@@ -0,0 +1,1241 @@
+      _server_monitor_heartbeat_failed (
+         server_monitor, &description->error, duration_us, awaited);
+   } else {
+      MONITOR_LOG_ERROR (
+         server_monitor, "network error occurred: %s", error.message);
+      command_or_network_error = true;
+      mongoc_server_description_handle_ismaster (description, NULL, -1, &error);
+      _server_monitor_heartbeat_failed (
+         server_monitor, &description->error, duration_us, awaited);
+   }
+
+   if (command_or_network_error) {
+      if (server_monitor->stream) {
+         mongoc_stream_failed (server_monitor->stream);
+      }
+      server_monitor->stream = NULL;
+      server_monitor->more_to_come = false;
+      bson_mutex_lock (&server_monitor->topology->mutex);
+      _mongoc_topology_clear_connection_pool (server_monitor->topology,
+                                              server_monitor->description->id);
+      bson_mutex_unlock (&server_monitor->topology->mutex);
+   }
+
+   bson_destroy (&ismaster_reply);
+   return description;
+}
+
+/* Request scan of a single server.
+ *
+ * Caller does not need to have topology mutex locked.
+ * Locks server monitor mutex to deliver scan_requested.
+ */
+void
+mongoc_server_monitor_request_scan (mongoc_server_monitor_t *server_monitor)
+{
+   MONITOR_LOG (server_monitor, "requesting scan");
+   bson_mutex_lock (&server_monitor->shared.mutex);
+   server_monitor->shared.scan_requested = true;
+   mongoc_cond_signal (&server_monitor->shared.cond);
+   bson_mutex_unlock (&server_monitor->shared.mutex);
+}
+
+/* Request cancellation of an in progress awaitable ismaster.
+ *
+ * Called from app threads on network errors and during shutdown.
+ * Caller must not hold lock.
+ * Locks server monitor mutex.
+ */
+void
+mongoc_server_monitor_request_cancel (mongoc_server_monitor_t *server_monitor)
+{
+   MONITOR_LOG (server_monitor, "requesting cancel");
+   bson_mutex_lock (&server_monitor->shared.mutex);
+   server_monitor->shared.cancel_requested = true;
+   mongoc_cond_signal (&server_monitor->shared.cond);
+   bson_mutex_unlock (&server_monitor->shared.mutex);
+}
+
+/* Wait for heartbeatFrequencyMS or minHeartbeatFrequencyMS if a scan is
+ * requested.
+ *
+ * Locks server monitor mutex.
+ */
+void
+mongoc_server_monitor_wait (mongoc_server_monitor_t *server_monitor)
+{
+   int64_t start_ms;
+   int64_t now_ms;
+   int64_t scan_due_ms;
+
+   start_ms = _now_ms ();
+   scan_due_ms = start_ms + server_monitor->heartbeat_frequency_ms;
+
+   bson_mutex_lock (&server_monitor->shared.mutex);
+   while (true) {
+      int64_t sleep_duration_ms;
+
+      if (server_monitor->shared.state != MONGOC_THREAD_RUNNING) {
+         break;
+      }
+
+      now_ms = _now_ms ();
+      if (now_ms >= scan_due_ms) {
+         break;
+      }
+
+      if (server_monitor->shared.scan_requested) {
+         server_monitor->shared.scan_requested = false;
+         scan_due_ms = start_ms + server_monitor->min_heartbeat_frequency_ms;
```

Comment:
If we do successfully sleep below, we'll jump to a new iteration of the `while` loop. Assuming we don't break because (a) server monitor thread is no longer running and (b) `now_ms` has not surpassed `scan_due_ms`, then are we sure that `scan_due_ms` is getting assigned correctly? `start_ms + server_monitor->min_heartbeat_frequency_ms` will not have changed since the last iteration.

Should this be calculated using `now_ms` instead, such that `start_ms` is only used for the initial `scan_due_ms` calculation (before the `while` loop)?

Reply:
Ah, I think I was reading this all backwards. If the `while` loop repeats, that means `mongoc_cond_timedwait` didn't sleep for as long as we need to because something woke up the thread. That's why we abort this loop and return from the function if either `sleep_duration_ms` is negative (no waiting required) or `mongo_cond_ret_is_timedout` is true (we waited the full duration we needed to).

This LGTM.